payments/payment_gateways/doctype/paymob_settings/paymob_settings.py

```diff
@@ -0,0 +1,266 @@
+			return "Invalid or missing data"
+
+		# Validate the HMAC
+		validator = HMACValidator(
+			incoming_hmac=incoming_hmac,
+			callback_dict=incoming_data_json
+		)
+
+		if not validator.is_valid:
+			return "Invalid HMAC"
+
+		# HMAC is valid, extract transaction data
+		obj_data = incoming_data_json.get("obj", {})
+		success = obj_data.get("success")
+		pending = obj_data.get("pending")
+		payment_status = obj_data.get("order", {}).get("payment_status")
+		txn_response_code = obj_data.get("data", {}).get("txn_response_code")
+		paymob_payment_id = obj_data.get("id")
+		paymob_order_id = obj_data.get("order", {}).get("id")
+
+		# Validate all success conditions
+		is_payment_successful = (
+			success is True and
+			pending is False and
+			str(payment_status).upper() == "PAID" and
+			str(txn_response_code).upper() == "APPROVED"
+		)
+
+		# Find the Integration Request based on order_id
+		if not paymob_order_id:
+			return "Missing order ID"
+
+		integration_requests = frappe.get_all(
+			"Integration Request",
+			filters={
+				"integration_request_service": "Paymob",
+				"data": ["like", f'%\"paymob_order_id\": \"{paymob_order_id}\"%']
+			},
+			fields=["name", "data", "reference_doctype", "reference_docname"],
+			order_by="creation desc",
+			limit=1
+		)
+
+		import json
+		integration_request_doc = frappe.get_doc("Integration Request", integration_requests[0].name)
+		integration_request_dict = json.loads(integration_request_doc.data)
+		integration_request_dict["paymob_payment_id"] = str(paymob_payment_id)
+		integration_request_dict["order_id"] = str(paymob_order_id)
+		
+		# Update the data field with additional information
+		integration_request_doc.data = json.dumps(integration_request_dict)
+		
+		# Check if transaction succeeded / paid
+		if is_payment_successful:
+			# Update integration request status
+			integration_request_doc.save(ignore_permissions=True)
+			frappe.db.commit()
+			
+			# Call on_payment_authorized on the reference document
+			if integration_request_dict['reference_doctype'] and integration_request_dict['reference_docname']:
+				custom_redirect_to = None
+				try:
+					custom_redirect_to = frappe.get_doc(
+						integration_request_dict['reference_doctype'],
+						integration_request_dict['reference_docname']
+					).run_method("on_payment_authorized", "Completed")
+					
+
+				except Exception:
+					frappe.log_error(frappe.get_traceback(), "Paymob on_payment_authorized Error")
+
+				if custom_redirect_to:
+					frappe.local.response["type"] = "redirect"
+					frappe.local.response["location"] = custom_redirect_to
+					return
+
+			return "Payment verified successfully"
+		else:
+			# Payment failed or was cancelled
+			integration_request_doc.status = "Failed"
+			integration_request_doc.error = f"Payment Status: {payment_status}, Response Code: {txn_response_code}"
+			integration_request_doc.save(ignore_permissions=True)
+			frappe.db.commit()
+			
+			return "Payment failed or was cancelled"
```

Comment:
Again why are you returning strings in this function? A success response in any case?